DataPerProcessing/utilities.py

```diff
@@ -90,27 +124,81 @@ def polygonToRotRectangle(bbox):
     h = ymax - ymin + 1
-
     return [float(center[0]),float(center[1]),w,h,angle]
 
-#print(polygonToRotRectangle(getBBoxList(labelPath+os.sep+'P0000.txt')[0]))
\ No newline at end of file
+#print(polygonToRotRectangle(getBBoxList(labelPath+os.sep+'P0000.txt')[0]))
+
+#------------------------------------------------------------------------
+# Function Name: cropImageBBox()
+# Description: Crop the object and make bg black
+#------------------------------------------------------------------------
+def cropImageBBox(imagePath, bbox, mode='show', outPath=None):
+	'''
+	* imagePath: raw image path from data set
+	* bbox     : The polygon stored in format [x1, y1, x2, y2, x3, y3, x4, y4]
+	* mode     : 'show' or 'save'
+	* outPath  : if mode is 'save' then output path
+	'''
+	x1, y1, x2, y2, x3, y3, x4, y4 = bbox
+	image = cv2.imread(imagePath)
+	mask = np.zeros(image.shape, dtype=np.uint8)
+	# mask defaulting to black for 3-channel and transparent for 4-channel
+	# (of course replace corners with yours)
+	roi_corners = np.array([[(x1,y1), (x2,y2), (x3,y3), (x4,y4)]], dtype=np.int32)
+	# fill the ROI so it doesn't get wiped out when the mask is applied
+	channel_count = image.shape[2]  # i.e. 3 or 4 depending on your image
+	ignore_mask_color = (255,)*channel_count
+	cv2.fillPoly(mask, roi_corners, ignore_mask_color)
+	# from Masterfool: use cv2.fillConvexPoly if you know it's convex
+
+	# apply the mask
+	masked_image = cv2.bitwise_and(image, mask)
+	xmin = min(x1,x2,x3,x4)
+	xmax = max(x1,x2,x3,x4)
+	ymin = min(y1,y2,y3,y4)
+	ymax = max(y1,y2,y3,y4)
+	image = masked_image[ymin:ymax, xmin:xmax] 
+	# save the result
+	if mode == 'show':
+		cv2.imshow('Image',image)
+		cv2.waitKey()
+	elif mode == 'save':
+		if outPath is not None:
+			cv2.imwrite(outPath, image)
+		else:
+			print('No Output path Provided')
+
+#------------------------------------------------------------------------
+# Function Name: sendToCropImageBBox()
+# Description: Preprocessing before sending to cropImageBBox
+#------------------------------------------------------------------------
+def sendToCropImageBBox(mainPath):
+	'''
+	* mainPath: Path where sub-folder with object class name will create
+	'''
+	ObjectCoordinatesList, ObjectNamesList, ObjectDifficultyList = getBBoxList(labelPath+os.sep+labelName)
+	print('Number of Objects: {}'.format(len(ObjectCoordinatesList)))
+	for dirName in set(ObjectNamesList):
+		if not os.path.exists(mainPath+os.sep+dirName):
+	    			os.makedirs(mainPath+os.sep+dirName)
+	printProgressBar(0, len(ObjectCoordinatesList), prefix = 'Progress:', suffix = 'Complete', length = 80)
+	for i in range(len(ObjectCoordinatesList)):
+		printProgressBar(i + 1, len(ObjectCoordinatesList), prefix = 'Progress:', suffix = 'Complete', length = 80)
+		path = mainPath+os.sep+ObjectNamesList[i]+os.sep+ObjectNamesList[i]+'_'+str(i)+'.png'
+		cropImageBBox(imagePath+os.sep+imageName, ObjectCoordinatesList[i], 'save', path)
+
+#sendToCropImageBBox('Objects')
```

Comment:
wherever you are using + operator for concatenation, you should use % operator for better memory management because every time a new string is created, python memory manager creates a new string and new address as string is immutable object, so use % to decrease the space complexity.